### PR TITLE
Update message and template data model

### PR DESCRIPTION
**Pull request by Thurfjell**
### Ticket:

- [15259](https://dev.azure.com/OGCIO-Digital-Services/Digital%20Services%20Programme/_backlogs/backlog/Life%20Events/Stories?workitem=15259)

### Description

Updated message and template model slightly. 
Updated function calls to the new
Added option to use template when creating a message

## Type

- [ ] **Dependency upgrade**
- [ ] **Bug fix**
- [x] **New feature**
- [x] **Dev change**

### Checklist:

- [ ] I have added tests that prove my fix is effective or that my feature works

### Screenshots:

N/A

### CODE REVIEW

File: apps/messaging/app/[locale]/admin/send-a-message/[action]/ComposeMessageMeta.tsx [hunk 1]
```diff
@@ -114,7 +122,10 @@ export default (props: MessageCreateProps) => {
               type="radio"
               value="message"
               className="govie-radios__input"
-              defaultChecked
+              defaultChecked={
+                !props.state.messageType ||
+                props.state.messageType === "message"
```
Comment (SamSalvatico) [inline apps/messaging/app/[locale]/admin/send-a-message/[action]/ComposeMessageMeta.tsx:127]: could we use an enum to define the message types?

File: apps/messaging/app/[locale]/admin/send-a-message/[action]/TemplateForm.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      <label className="govie-label--s">Plain text</label>
+      <p className="govie-body">{template?.plainText}</p>
+
+      <hr />
+
+      <h3>
+        <span className="govie-heading-s">Variables</span>
+      </h3>
+      <form action={action}>
+        {template.fields?.map((field) => (
```
Comment (SamSalvatico) [inline apps/messaging/app/[locale]/admin/send-a-message/[action]/TemplateForm.tsx:78]: `template?.fields?.map`

File: apps/messaging/app/[locale]/admin/settings/emails/provider/page.tsx [hunk 3]
```diff
@@ -83,12 +91,20 @@ export default async (props: { searchParams: { id: string } }) => {
...
 
     // Just for ts type assertions, at this point this is always false.
-    if (!name || !host || !username || !password || !port) {
+    if (!name || !host || !username || !password || !port || !fromAddress) {
       return;
     }
 
     if (!id) {
-      await mailApi.createProvider({ name, host, username, password, port });
+      await mailApi.createProvider({
```
Comment (SamSalvatico) [inline apps/messaging/app/[locale]/admin/settings/emails/provider/page.tsx:99]: is `throttle` managed in constructor?

Comment (Thurfjell) [reply apps/messaging/app/[locale]/admin/settings/emails/provider/page.tsx:99]: No it'll use whatever standards we conclude when this becomes a rest api.

File: apps/messaging/app/[locale]/admin/send-a-message/[action]/ComposeMessageMeta.tsx [hunk 4]
```diff
@@ -1,20 +1,21 @@
...
-import { MessageCreateProps } from "../../../../utils/messaging";
-import { useTranslations } from "next-intl";
+import { MessageCreateProps, MessageType } from "../../../../utils/messaging";
+import { getTranslations } from "next-intl/server";
+import { headers } from "next/headers";
 
-export default (props: MessageCreateProps) => {
-  const t = useTranslations("sendAMessage.ComposeMessageMeta");
+export default async (props: MessageCreateProps) => {
+  console.log(JSON.stringify(headers()));
```
Comment (GiuliaTeggi) [inline apps/messaging/app/[locale]/admin/send-a-message/[action]/ComposeMessageMeta.tsx:9]:
```suggestion
```

File: apps/messaging/app/[locale]/admin/send-a-message/[action]/TemplateForm.tsx [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        (acc, pair) => {
+          acc[pair.key] = pair.value;
+          return acc;
+        },
+        {},
+      ),
+      submittedContentAt: new Date().toISOString(),
+    };
+    const next = Object.assign({}, props.state, update);
+    console.log(next);
```
Comment (GiuliaTeggi) [inline apps/messaging/app/[locale]/admin/send-a-message/[action]/TemplateForm.tsx:34]:
```suggestion
```

File: apps/messaging/app/[locale]/messages/page.tsx [hunk 6]
```diff
@@ -8,14 +8,11 @@ import ds from "design-system";
...
-  const { email } = await PgSessions.get();
+  const { email, userId } = await PgSessions.get();
   const params = new URLSearchParams(props.searchParams);
 
-  const messages = await api.getMessages(email, {
-    search: params.get("search")?.toString() || "",
-    page: 1,
-    size: 10,
-  });
+  console.log({ userId });
```
Comment (GiuliaTeggi) [inline apps/messaging/app/[locale]/messages/page.tsx:14]:
```suggestion
```

Comment (GiuliaTeggi) [reply apps/messaging/app/[locale]/messages/page.tsx:14]: We really need eslint 🙈

File: packages/messages/index.ts [hunk 7]
```diff
@@ -110,20 +148,26 @@ export const mailApi = {
...
           user: username,
           pass: password,
         },
       });
+      console.log("Vi forsoker skicka mail", {
+        fromAddress,
+        recipients,
+        subject,
+        body,
+      });
```
Comment (GiuliaTeggi) [inline packages/messages/index.ts:167]: 🇸🇪

File: packages/messages/index.ts [hunk 8]
```diff
@@ -556,19 +276,608 @@ export const api = {
...
-          SELECT template_id, jsonb_agg(jsonb_build_object('key', key_accessor, 'value', value_accessor)) as keyvals
-          from template_interpolation_accessors
-          group by template_id
-        ) values on values.template_id = template.id;
-    `,
-      )
-      .then((res) => res.rows);
+};
+
+export const utils = {
```
Comment (GiuliaTeggi) [inline packages/messages/index.ts:281]: Can't you move this to a utils file?

File: apps/messaging/app/[locale]/messages/page.tsx (hunk 6 above)
Comment (Thurfjell) [reply apps/messaging/app/[locale]/messages/page.tsx:14]: YES

File: packages/auth/sessions.ts [hunk 9]
```diff
@@ -134,3 +134,28 @@ export async function getUserInfoById(userId: string) {
...
+  return pgpool
+    .query<{ id: string; email: string; lang: string }>(
+      `
+    SELECT id, govid_email as "email", 'en' as "lang" FROM users where is_public_servant = FALSE
+  `,
+    )
+    .then((res) => res.rows);
+}
+
+export async function getUsersFor(ids: string[]) {
```
[the remark this answers is not in the document]
Comment (Thurfjell) [reply packages/auth/sessions.ts:149]: Hahaha yep! 😂

File: packages/messages/index.ts (hunk 8 above)
Comment (Thurfjell) [reply packages/messages/index.ts:281]: Yep, I'll do it in the rest api PR.  😞

Comment (Thurfjell) [reply packages/messages/index.ts:281]: Actually, I think we might want both this package and the rest. I'll think of something.

File: packages/messages/index.ts (hunk 7 above)
Comment (Thurfjell) [inline packages/messages/index.ts:167]:
```suggestion
```

**Review by GiuliaTeggi — APPROVED** [review]

File: packages/messages/migrations/sqls/20240408095916-message-model-update-up.sql [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    plain_text text not null,
+    updated_at timestamptz,
+    created_at timestamptz not null default now(),
+    primary key(template_meta_id, lang)
+);
+
+create table message_template_variables (
+    template_meta_id uuid not null,
+    field_name text not null,
+    field_type text not null
```
Comment (Thurfjell) [inline packages/messages/migrations/sqls/20240408095916-message-model-update-up.sql:59]:
```suggestion
    field_type text not null,
    primary key(template_meta_id, field_name)
```

**Pull request merged**